sound/soc/sof/ipc.c

```diff
@@ -279,10 +283,7 @@ int sof_ipc_tx_message_unlocked(struct snd_sof_ipc *ipc, u32 header,
 	ipc_log_header(sdev->dev, "ipc tx", msg->header);
 
 	/* now wait for completion */
-	if (!ret)
-		ret = tx_wait_done(ipc, msg, reply_data);
-
-	return ret;
+	return tx_wait_done(ipc, msg, reply_data);
```

Comment:
@plbossart with the previous code you used to sleep in all cases. Look at the (old) code above that change. Before `snd_sof_dsp_send_msg()` could only return 0 or a negative error code. Then comes a check `if (ret < 0) {...; return ret;}` - so, if you don't return there, `ret = 0`. So, the below test `if (!ret)` is totally redundant.
Now I also added a case of `ret > 0`. And still, even with this change `tx_wait_done()` can be called always.

Reply:
I just don't get it, sorry...

One one side you say we don't need to wait but on the other now unconditionally calls tx_wait_done()

I am not able to reconcile those two opposite logical assertions. Either the comment is not worded correctly, or you assume a behavior in tx_wait_done() that is not explained?

Reply:
@plbossart ok, very simple. Here's what it was before (a bit simplified):
```
ret = snd_sof_dsp_send_msg(...);
if (ret < 0) {
	printk();
	return ret;
}
if (!ret)
	ret = tx_wait_done(ipc, msg, reply_data);
return ret;
```
Now, back then `snd_sof_dsp_send_msg()` could only return 0 or a negative error code. So, the above code can be trivially and equivalently reduced to
```
ret = snd_sof_dsp_send_msg(...);
if (ret < 0) {
	printk();
	return ret;
}
ret = tx_wait_done(ipc, msg, reply_data);
return ret;
```
Or one step further
```
ret = snd_sof_dsp_send_msg(...);
if (ret < 0) {
	printk();
	return ret;
}
return tx_wait_done(ipc, msg, reply_data);
```
That's it. This is equivalent to the original code before this patch.
Next, I add a case of `ret > 0` and it has to be handled specially, namely, there should be no waiting in that case. But this is already handled by virtue of
```
if (!ret)
	msg->ipc_complete = false;
```
Because `if (ret > 0)` the above won't be executed and `msg->ipc_complete` will remain at its default value of `true` and `tx_wait_done()` returns immediately without waiting in this case because of `wait_event_timeout(..., msg->ipc_complete,...)` - the event is already true, so, no waiting.

Reply:
Wow, that was a long one not easy on my shrinking pool of neurons. I am afraid the logical flow cannot be verified just by looking at the diff only, but only with extra context.
Here you call a tx_wait_done() function in all cases, but you didn't add a comment or explain that it returns immediately.

I respect your talent for optimization but if they take too much time to reverse-engineer you need to help the rest of us with hints on what is happening, e.g. with comments. Thanks!